src/installLogsPrinter.js

```diff
@@ -68,17 +72,25 @@ function installLogsPrinter(on, options = {}) {
       }
 
       if (options.outputTarget) {
+        allMessages[0] = data.state;
         allMessages[data.spec] = allMessages[data.spec] || {};
         allMessages[data.spec][data.test] = messages;
       }
 
-      logToTerminal(messages, options);
+      if ((options.printLogsToConsole === "onFail" && data.state !== "passed")
+        || options.printLogsToConsole === "always"){
+        logToTerminal(messages, options);
+      }
+
       return null;
     },
     [CONSTANTS.TASK_NAME_OUTPUT]: () => {
       outputProcessors.forEach((processor) => {
-        processor.write(allMessages);
-        logOutputTarget(processor);
+        if ((options.printLogsToFile === "onFail" && allMessages[0] !== "passed")
```

Comment:
You will have to go through each spec here and filter out the ones that suceeded in case it's set to onFail.

Reply:
What if I only collect those messages when I need to?
```
if (options.outputTarget && options.printLogsToFile !== "never") {
        if (data.state === "failed" || options.printLogsToFile === "always") {
          allMessages[data.spec] = allMessages[data.spec] || {};
          allMessages[data.spec][data.test] = messages;
        }
      }
```